jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/impl/AbstractAvailableJobsExecutor.java

```diff
@@ -0,0 +1,255 @@
+                    Thread.currentThread().interrupt();
+                } catch (Throwable e) {
+                	callbacks = classCacheManager.buildCommandCallback(ctx, cl);  
+                    logger.warn("Error during command {} error message {}", request.getCommandName(), e.getMessage(), e);
+    
+                    ErrorInfo errorInfo = new ErrorInfo(e.getMessage(), ExceptionUtils.getStackTrace(e.fillInStackTrace()));
+                    errorInfo.setRequestInfo(request);
+    
+                    ((List<ErrorInfo>)request.getErrorInfo()).add(errorInfo);
+                    logger.debug("Error Number: {}", request.getErrorInfo().size());
+                    if (request.getRetries() > 0) {
+                        request.setStatus(STATUS.RETRYING);
+                        request.setRetries(request.getRetries() - 1);
+                        
+                        
+                        // calculate next retry time
+                        List<Long> retryDelay = (List<Long>) ctx.getData("retryDelay");
+                        if (retryDelay != null) {
+                            long retryAdd = 0l;
+
+                            try {
+                                retryAdd = retryDelay.get(request.getExecutions());
+                            } catch (IndexOutOfBoundsException ex) {
+                                // in case there is no element matching given execution, use last one
+                                retryAdd = retryDelay.get(retryDelay.size()-1);
+                            }
+                            
+                            request.setTime(new Date(System.currentTimeMillis() + retryAdd));
+                            request.setExecutions(request.getExecutions() + 1);
+                            logger.info("Retrying request ( with id {}) - delay configured, next retry at {}", request.getId(), request.getTime());
+                        }
+                        
+                        logger.debug("Retrying ({}) still available!", request.getRetries());
+                        
+                        executorStoreService.updateRequest(request);
+                    } else {
+                        logger.debug("Error no retries left!");
+                        request.setStatus(STATUS.ERROR);
+                        request.setExecutions(request.getExecutions() + 1);
+                        
+                        executorStoreService.updateRequest(request);
+                        
+                        if (callbacks != null) {
+                            for (CommandCallback handler : callbacks) {                        
+                                handler.onCommandError(ctx, e);                        
+                            }
+                        }
+                        processReoccurring = true;
+                    }
+    
+                } finally {
+                	if (processReoccurring && cmd != null && cmd instanceof Reoccurring) {
```

Comment:
you mean to put all code from catch and finally into own methods? If so that's actually good idea, will make that happen :)